### PR TITLE
Enhance and complement OPcache setup checks

### LINKED ISSUE #15522: Do OPcache recommendations based on actual cache usage
**Is your feature request related to a problem? Please describe.**
- Currently OPcache settings (`opcache.memory_consumption`, `opcache.interned_strings_buffer`, `opcache.max_accelerated_files`) recommendations on Nextcloud admin panel are fixed values that match the current PHP defaults: https://php.net/manual/en/opcache.configuration.php
- For Nextcloud-only servers, based on tests with all available apps installed, those are never reached and could be halved.
- If other large PHP sites are present, those could be too low.

**Describe the solution you'd like**
- It makes sense to prompt recommendations for each setting separately, based on the actual OPcache usage.
- The [opcache_get_status()](https://www.php.net/manual/en/function.opcache-get-status.php) function could be used for this. It could then be recommended to raise a value if e.g. more than 80% of the limit is reached.

**Describe alternatives you've considered**
- Having failsafe high recommendations (current implementation) does not hurt much either, especially since there seems to be no negative performance impact when having "too" high values.
- The main benefit of the requested change is then that most users (with default OPcache values active) will not face any warning while in rare cases, where higher values are required, those are covered by the warning(s) as well. Also shared hosting users, that have no access to the PHP settings, will not be bothered by the warnings, or, can do requests to their hosting provider based on a real measure.

**Additional context**
Related discussion has been done here: https://github.com/nextcloud/documentation/issues/1439

### LINKED ISSUE #27094: Make "The PHP OPcache is not properly configured" a little bit more tolerant and precise (Suggestion)
Related Code: `apps/settings/lib/Controller/CheckSetupController.php: protected function isOpcacheProperlySetup()`
Related Doc: https://docs.nextcloud.com/server/21/admin_manual/installation/server_tuning.html#enable-php-opcache

---
After monitoring our OpCaches, the function `isOpcacheProperlySetup()` should be a bit more tolerant and precise. These are our suggestions:

**max_accelerated_files**
```
		if ($this->iniGetWrapper->getNumeric('opcache.max_accelerated_files') < 10000) {
			return false;
		}
```
According to php.net:
> The maximum number of keys (and therefore scripts) in the OPcache hash table. The actual value used will be the first number in the set of prime numbers `{ 223, 463, 983, 1979, 3907, 7963, 16229, 32531, 65407, 130987, 262237, 524521, 1048793 }` that is greater than or equal to the configured value. The minimum value is 200. The maximum value is 1000000. Values outside of this range are clamped to the permissible range. 

Means: As long as Nextcloud consists of fewer than 16229 files, check on `if ($this->iniGetWrapper->getNumeric('opcache.max_accelerated_files') <= 7963)` (some Admins like to configure the precise values, and 10000 is as good as 7964).

---

**memory_consumption**
```
		if ($this->iniGetWrapper->getNumeric('opcache.memory_consumption') < 128) {
			return false;
		}
```
Nextcloud instances with 30 to 120 active (!) users use between 40 and 60 mb of opcache, so we would suggest that a check for 80 mb is sufficient: `if ($this->iniGetWrapper->getNumeric('opcache.memory_consumption') < 80) {`

---

**interned_strings_buffer**
```
		if ($this->iniGetWrapper->getNumeric('opcache.interned_strings_buffer') < 8) {
			return false;
		}
```
8 MB memory used to store interned strings is not enough (if you configure "8", you get 6 MB Buffer Size). 16 MB seem to be enough (you will get 12 MB Buffer Size): `if ($this->iniGetWrapper->getNumeric('opcache.interned_strings_buffer') < 16) {`

---

**Doc**
It could be emphasized that `opcache.save_comments=1` is a must, because disabling this configuration directive breaks Nextcloud because it relys on comment parsing for annotations.

### PR DESCRIPTION
**Pull request by MichaIng**
The current OPcache recommendations match the PHP defaults, but the values are much higher than required to run Nextcloud, even with a high number of installed apps. On the other hand, when other applications use the same OPcache instance, the recommended values might not be sufficient. Accurate recommendations need to take into account actual OPcache usage.

With this commit, recommendations are shown to raise the config value if more than 90% of max cache size or number of keys is used.

The checks whether the module is loaded and whether the OPcache is properly configured have been merged into a single function. This allowed to reduce the overhead of OPcache configuration checks when the module is not loaded.

A check has been added whether Nextcloud is permitted to use the OPcache API. Without this, inconsistencies during core or app upgrades may cause errors and OPcache usage cannot be determined for the new usage based checks.

OPcache usage based checks are skipped when Nextcloud is not permitted to use the API, or when the `opcache_get_status` function has been explicitly disabled: [see below](#issuecomment-926115962)

### CODE REVIEW

**Comment by MichaIng** [comment]
php-cs check reports:
```
Warning: Found violation(s) of type: class_definition
Warning: Found violation(s) of type: braces
```
Not sure if this is related to this commit, probably someone can give me a hint what is wrong?

Then there is the `PossiblyNullOperand` error, as I do:
```
$recommendations[] = 'text' . $this->iniGetWrapper->getNumeric('opcache.max_accelerated_files') . 'text';
```
If OPcache is enabled, this cannot be null, but still good to rule it out code-wise, I guess. Is there a way to do that in the same line, or do I need to assign the result in a variable, and check it in a separate line?

All other errors/warnings are not related to this PR, as far as I can see.

**Comment by MichaIng** [comment]
Functional tests passed:
opcache

Probably the output lines can be a little enhanced, e.g. showing config key and value on a separate line, or making them fat?

Also we could recommend the next power of two the interned strings buffer, as smaller increases can lead to even a reduced resulting cache size: https://stackoverflow.com/q/67853338

In case of max accelerated files we could recommend the next effective value, as e.g. raising the config from **10000** to **16000** has zero effect, as in both cases **16229** is the resulting value: https://www.php.net/manual/opcache.configuration.php#ini.opcache.max-accelerated-files

But too much magic in the check is probably not great and instead it could make more sense to mention such details in the docs, and re-add the docs URL for further details. The docs should be adjusted then as well, I'll open a companion PR when this one is generally accepted.

File: apps/settings/lib/Controller/CheckSetupController.php [hunk 1]
```diff
@@ -426,31 +426,39 @@ public function getFailedIntegrityCheckFiles(): DataDisplayResponse {
...
-			return false;
-		}
+			// Check for saved comments only when OPcache is currently disabled. If it was enabled, opcache.save_comments=0 would break Nextcloud in the first place.
+			if (!$this->iniGetWrapper->getBool('opcache.save_comments')) {
+				$recommendations[] = 'OPcache is configured to remove code comments. With OPcache enabled, <code>opcache.save_comments=1</code> must be set for Nextcloud to function.';
+			}
+		} else {
+			// Mute error when opcache.restrict_api is set and does not permit Nextcloud to use opcache_get_status(). All below conditions will then return false.
+			// ToDo: Add a check for opcache.restrict_api, since it can cause issues when Nextcloud cannot evict cached scripts: https://github.com/nextcloud/server/pull/8188
+			$status = @opcache_get_status();
```
Comment (kesselb) [inline apps/settings/lib/Controller/CheckSetupController.php:445]:
```suggestion
			$status = @opcache_get_status(false);
```

**Comment by kesselb** [comment]
Thanks for your pull request :+1: 

Code looks good. The recommendations make sense to me. In serverinfo we have an additional check for the module `extension_loaded('Zend OPcache')`.

I prefer https://github.com/nextcloud/server/pull/27301 because it does not call opcache_get_status and just check for the minimum values. We should have a short section in our documentation that opcache needs to adjusted for the actual use case and give some examples and references to the php documentation. 

We have to take into account the people hate the setup checks. Especially when they are not able to fix it. 
It's (for some) super important to see the green check in the admin overview and we have many discussions on GitHub and help.nextcloud.com about that. 

What do you think about writing an app for Nextcloud to monitor the opcache usage, give recommendations how to optimize and maybe take over the opcache api from serverinfo?

File: apps/settings/lib/Controller/CheckSetupController.php (hunk 1 above)
Comment (MichaIng) [reply apps/settings/lib/Controller/CheckSetupController.php:445]: Ah nice, this makes totally sense. I was concerned as well that the whole list of cached keys is returned.

**Comment by MichaIng** [comment]
Whoops, sorry for the long text 😄...

> In serverinfo we have an additional check for the module `extension_loaded('Zend OPcache')`.

You mean `hasOpcacheLoaded()`? Yes that makes sense and is untouched by this commit. `isOpcacheProperlySetup` is only evaluated if the module itself is enabled. But actually, both functions are always called, while `isOpcacheProperlySetup()` could be skipped in the first place when `hasOpcacheLoaded()` returned false, to reduce overhead a little. Not sure how to achieve this best with the way the `DataResponse` is currently generated. Both functions could be merged via:
```
protected function isOpcacheProperlySetup(): array {
	$recommendations = [];

	if (!extension_loaded('Zend OPcache')) {
			$recommendations = ['The PHP OPcache module is not loaded. <a target="_blank" rel="noreferrer noopener" class="external" href="' . $this->urlGenerator->linkToDocs('admin-php-opcache') . '">For better performance it is recommended ↗</a> to load it into your PHP installation.'];
	} elseif (!$this->iniGetWrapper->getBool('opcache.enable')) {
...
```
which then also simplifies the evaluation in `setupchecks.js`.

> I prefer #27301 because it does not call opcache_get_status and just check for the minimum values.

`#27301` does not address the underlying issue:
- It lowers the `opcache.max_accelerated_files` to `7963`, to match one of the actually effective prime numbers. Not sure why PHP defaults to the ineffective `10000`, which always results `16229` max keys. But there is no guarantee that this is sufficient or still too high for the particular case. In my case ~2000 files are cached, so `3907` is sufficient. When you run other web applications on the same server, even `16229` might be too low.
- It **raises** the recommendation for `opcache.interned_strings_buffer` from `8` to `16`, as in the authors particular case, `8` was not sufficient. In my case ~3.5 MiB interned strings buffer is used, with an older PHP version it was 2-3 MiB, hence `8` is sufficient (`4` leads to `3145304` bytes, which was sufficient earlier), but in other cases again it can be too low or too high. The interned strings buffer changed quite a lot with the last PHP versions, so a fixed value cannot be right at all currently.
- `opcache.memory_consumption` is left untouched. I once tried to install literally ALL apps from the Nextcloud apps store (not all were possible, since some conflicted or did not work well with current NC that time) and accessed all their settings and apps pages, so test how high the OPcache usage can raise. And I was reaching only a tiny little above 64 MiB. This cannot be seen as representative of course. My current instance uses ~30 MiB OPcache memory (which includes the interned strings buffer always). But also here, as my instance and the authors comment that "some mentions of specific apps capable of throwing the memory requirement" shows quite well that it is very case-specific (use case, installed apps, PHP version, ...) and it cannot be right to recommend a fixed set of values.

> We should have a short section in our documentation that opcache needs to adjusted for the actual use case and give some examples and references to the php documentation.

Yes, definitely, and this is what I'll do when this PR is accepted. But most users do not read every section of the docs, I suppose, and OPcache is critical for access performance, so doing as good as possible recommendations, that match the actual usage on the particular system, IMO is important. Otherwise users might complain about bad Nextcloud performance (compared to other applications with less OPcache usage), while raising max buffer values could fix it.

> We have to take into account the people hate the setup checks. Especially when they are not able to fix it.

Exactly, we regularly receive reports of users who want to have all admin panel recommendations/warnings resolved, regardless if it can be sometimes well argued that they can/should be ignored in a particular case. In most cases the recommended values are too high currently, so this change will reduce the warnings in mos cases. If the usage indeed exceeds the currently applied value too often, e.g. because PHP internally handles those differently or Nextcloud grows, it could be discussed whether to have the specifically affected value removed from the test and refer to the documentation instead. But for now I'm 100% sure that this PR helps reducing the warning. `#27094` clearly implies that the recommendations are shown much more often, as the interned strings buffer in PHP defaults to `8`, which is the current recommendation, but that PR raises it to `16`. So practically all users will at first see that warning (while currently none does when the values were not manually reduced) and I guess nearly all users of shared hostings as well, when each customer has an own OPcache instance, at least.

> What do you think about writing an app for Nextcloud to monitor the opcache usage, give recommendations how to optimize and maybe take over the opcache api from serverinfo?

If someone has time to write it, why not. But actually the change and check here is quite simple and preserves what the admin panel checks are intended for, including the IMO quite performance aspect, that is part of the checks headline for a reason. There are other checks which are IMO way less relevant or should be part of a discussion to have them removed and addressed differently instead, like the case of the Imagick module or [these](https://github.com/nextcloud/server/issues/22849). So when it's about reducing the amount of warnings, then I'd personally not start with OPcache. To have some more overview over OPcache usage in general, allow manual invalidation etc, what I aim to add to the docs is a hint about https://github.com/amnuts/opcache-gui.

**Comment by kesselb** [comment]
In addition to `opcache.restrict_api` it's also valid to add opcache_get_status to disable_functions: https://help.nextcloud.com/t/20-0-1-errors-phpstatistics-opcache-get-status-has-been-disabled-for-security-reasons/97231

`php -d opcache.enable_cli=1 -d disable_functions="opcache_get_status" -r "var_dump(opcache_get_status());"`

**Comment by MichaIng** [comment]
Thanks. Makes sense to checks this as well before calling it then. Since this only means the checks cannot be done, while it does not affect Nextcloud directly, shall we add a "recommendation" about `disable_functions="opcache_get_status"`, i.e. that it should be allowed for Nextcloud to be able to check OPcache usage, or shall we ignore it silently? There might be admins who want to have the NC admin panel free of warnings/recommendations but want to have `opcache_get_status` disabled, knowing/believing that their values are fine.

**Comment by MichaIng** [comment]
Btw, I don't think that the tests whether setting a certain flag/value in the array leads to a certain warning on admin panel is reasonable. We only test whether `setupchecks.js` parses/handles the passed array correctly, which does not involve any other code or logic, but we do not test whether an actual setup issue leads to the related warning (?).

Testing actual setup issues is definitely more difficult to implement, especially when it's about OPcache usage, but IMO it is worth it. Does someone have an idea where and how to do it best, e.g. a test which includes the PHP environment setup, the enabled modules and configs?

The topic should however be discussed in a separate issue.

**Comment by MichaIng** [comment]
Hmm, continuing my post above, now with `apps/settings/tests/Controller/CheckSetupControllerTest.php` we have another such type of tests, where I'd seriously question the whole purpose if it. Now we test in various ways whether within the same script an internal functions returns a value whether in the same script this value is then assigned to the intended variable. And what we had above already is whether if the variable is then set as expected, if the frontend then forges the correct HTML from it. Tons of code but whether the checks do successfully check what they are supposed to check is not covered at all.

Seriously, isn't it much better to setup a PHP instance with defined modules and settings, run the setup checks script on this PHP instance and check whether it prints warnings about the intentionally disabled or misconfigured modules? Two setups would be sufficient which contain for every check a warning in one case and no warning in the other case.

**Review by icewind1991 — APPROVED** [review]

**Review by CarlSchwan — APPROVED** [review]

**Comment by MichaIng** [comment]
I retested all combinations of OPcache and whether the warnings show up as expected, once on master and once on Nextcloud 23, which both works well.

**Comment by MichaIng** [comment]
Related documentation update is proposed here: https://github.com/nextcloud/documentation/pull/7859

**Pull request merged**